internal/integritymonitor/integritymonitor.go

```diff
@@ -0,0 +1,266 @@
+			log.Error(ctx.Err())
+			return ctx.Err()
+		case countHashes := <-comparedHashesChan:
+			log.WithField("countHashes", countHashes).Info("hashes compared successfully")
+			log.Debug("end check integrity")
+			return nil
+		case err := <-errC:
+			integrityCheckFailed(log, err, alertSender, kubeClient)
+		}
+	}
+}
+
+func saveHashes(
+	ctx context.Context,
+	hashC <-chan worker.FileHash,
+	dd *models.DeploymentData,
+	errC chan<- error,
+) <-chan int {
+	doneC := make(chan int)
+
+	go func() {
+		defer close(doneC)
+
+		const defaultHashCnt = 100
+		hashData := make([]*api.HashData, 0, defaultHashCnt)
+		alg := viper.GetString("algorithm")
+		countHashes := 0
+
+		for v := range hashC {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			hashData = append(hashData, fileHashToDtoDB(alg, &v))
+			countHashes++
+		}
+
+		ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
+		defer cancel()
+
+		query, args := data.NewHashFileData().PrepareBatchQuery(hashData, dd)
+		err := repositories.ExecQueryTx(ctx, query, args...)
+		if err != nil {
+			errC <- err
+		}
+		doneC <- countHashes
+	}()
+
+	return doneC
+}
+
+func compareHashes(
+	ctx context.Context,
+	log *logrus.Logger,
+	hashC <-chan worker.FileHash,
+	directory string,
+	algName string,
+	deploymentData *models.DeploymentData,
+	errC chan<- error,
+) <-chan int {
+	doneC := make(chan int)
+	go func() {
+		defer close(doneC)
+
+		repository := repositories.NewAppRepository(log, repositories.DB().SQL())
+		expectedHashes, err := repository.GetHashData(
+			directory,
+			algName,
+			deploymentData,
+		)
+		if err != nil {
+			errC <- fmt.Errorf("failed get hash data: %w", err)
+			return
+		}
+
+		//convert hashes to map
+		expectedHashesMap := make(map[string]string)
+		for _, h := range expectedHashes {
+			expectedHashesMap[h.FullFilePath] = h.Hash
+		}
+
+		for v := range hashC {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			if h, ok := expectedHashesMap[v.Path]; ok {
+				if h != v.Hash {
+					errC <- &IntegrityError{Type: ErrTypeFileMismatch, Path: v.Path, Hash: v.Hash}
+					return
+				}
+				delete(expectedHashesMap, v.Path)
+			} else {
+				errC <- &IntegrityError{Type: ErrTypeNewFile, Path: v.Path, Hash: v.Hash}
+				return
+			}
+		}
+		for p, h := range expectedHashesMap {
+			errC <- &IntegrityError{Type: ErrTypeFileDeleted, Path: p, Hash: h}
+			return
+		}
+		doneC <- len(expectedHashes)
+	}()
+	return doneC
+}
+
+func integrityCheckFailed(
+	log *logrus.Logger,
+	err error,
+	alertSender alerts.Sender,
+	kubeClient *services.KubeClient,
+) {
+
+	var msg string
+	var path string
+	var integrityError *IntegrityError
+	if errors.As(err, &integrityError) {
+		path = integrityError.Path
+		switch integrityError.Type {
+		case ErrTypeFileMismatch:
+			msg = IntegrityMessageFileMismatch
+		case ErrTypeNewFile:
+			msg = IntegrityMessageNewFileFound
+		case ErrTypeFileDeleted:
+			msg = IntegrityMessageFileDeleted
+		}
```

Comment:
func (e *IntegrityError) Error() {..}

Reply:
..and log this error on the upper layer.

Reply:
msg is text message for Splunk